server/services/timelineService.ts

```diff
@@ -1,28 +1,83 @@
+
+      const timeline = toTimeline(timelineResponse)
+      timeline.events = await this.addPrisonNameToPrisons(timeline.events, systemToken)
+      return timeline
     } catch (error) {
       logger.error(`Error retrieving Timeline for Prisoner [${prisonNumber}]: ${error}`)
       return { problemRetrievingData: true } as Timeline
     }
   }
 
-  filterTimelineByActionPlanCreatedOrUpdatedEvent = (timelineResponse: TimelineResponse) => {
-    return timelineResponse.events.filter(
-      (event: { eventType: string }) =>
-        event.eventType === 'ACTION_PLAN_CREATED' || event.eventType === 'INDUCTION_UPDATED',
-    )
+  private addPrisonNameToPrisons = async (
+    events: Array<TimelineEvent>,
+    token: string,
+  ): Promise<Array<TimelineEvent>> => {
+    if (!events || events.length === 0) {
+      return []
+    }
+
+    // Lookup the prison for the first event by itself to prevent a race condition on the `PrisonService`
+    // If the `PrisonStore` cache is empty/stale, ettempting to lookup all the event's prisons in a `Promise.all`
+    // loop results in a race condition where all prison look ups run (essentially) at the same time, and they all try
+    // to populate the cache from the `prison-register-api`. This results in many REST API calls to `prison-register-api`
+    // effectively spamming `prison-register-api` !
+    const firstEvent: TimelineEvent = {
+      ...events[0],
+      prison: await this.lookupPrison(events[0].prison.prisonId, token),
+    }
+
+    // Lookup the prisons for the remaining events. We can do this in an async loop, safe in the knowledge that the
+    // `PrisonService` is backed by a populated cache and therefore no race condition.
+    const otherEvents: Array<Promise<TimelineEvent>> = events.slice(1).map(async (event): Promise<TimelineEvent> => {
+      return {
+        ...event,
+        prison: await this.lookupPrison(event.prison.prisonId, token),
+      }
+    })
+
+    const eventPromises: Array<Promise<TimelineEvent>> = [
+      // Wrap the first event in a promise (it was created as a non-promise initially to force it to be created synchronously)
+      Promise.resolve().then(() => {
+        return firstEvent
+      }),
+      // The other events are already promises as they were created in an async loop
+      ...otherEvents,
+    ]
+    return Promise.all(eventPromises)
+  }
+
+  private async lookupPrison(prisonId: string, token: string): Promise<Prison> {
+    try {
+      return (await this.prisonService.getPrisonByPrisonId(prisonId, token)) || { prisonId, prisonName: undefined }
+    } catch (e) {
+      logger.error(`Error looking up prison ${prisonId}`, e)
+      // return a Prison with just the prison ID set. Failing to lookup the prison should not stop the Timeline service returning a Timeline
+      return { prisonId, prisonName: undefined }
+    }
   }
+
+  private filterTimelineEvents = (event: TimelineEventResponse): boolean =>
```

Comment:
I've changed the way we filter the events - this approach will be easier to maintain moving forwards as we can simply add to the `SUPPORTED_TIMELINE_EVENTS` array at the top of the file 👍